include/drivers/reset.h

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2022 Andrei-Edward Popa <andrei.popa105@yahoo.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public Reset Controller driver APIs
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RESET_H_
+#define ZEPHYR_INCLUDE_DRIVERS_RESET_H_
+
+/**
+ * @brief Reset Controller Interface
+ * @defgroup reset_controller_interface Reset Controller Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * API template to get the reset status of the device.
+ *
+ * @see reset_status
+ */
+typedef int (*reset_api_status)(const struct device *dev, uint16_t offset,
+								uint8_t regbit, uint8_t *status);
+
+/**
+ * API template to put the device in reset state.
+ *
+ * @see reset_assert
+ */
+typedef int (*reset_api_assert)(const struct device *dev, uint16_t offset, uint8_t regbit);
+
+/**
+ * API template to take out the device from reset state.
+ *
+ * @see reset_deassert
+ */
+typedef int (*reset_api_deassert)(const struct device *dev, uint16_t offset, uint8_t regbit);
+
+/**
+ * API template to reset the device.
+ *
+ * @see reset_reset
+ */
+typedef int (*reset_api_reset)(const struct device *dev, uint16_t offset, uint8_t regbit);
+
+/**
+ * @brief Reset Controller driver API
+ */
+__subsystem struct reset_driver_api {
+	reset_api_status status;
+	reset_api_assert assert;
+	reset_api_deassert deassert;
+	reset_api_reset reset;
```

Comment:
I'm a bit confused about the usage of `offset` and `regbit`. You are implicitly assuming that the reset controller is using some MMIO mapping (offset + regbit) but this could not be the case for every hardware.

For example Linux just uses a generic `unsigned long id` for the line.